overlord/devicestate/devicestate.go

```diff
@@ -1411,35 +1457,261 @@ func createRecoverySystemTasks(st *state.State, label string, snapSetupTasks []s
+		}
+	}
+
+	if err := valsets.Conflict(); err != nil {
+		return nil, err
+	}
+
+	revisions, err := valsets.Revisions()
+	if err != nil {
+		return nil, err
+	}
+
+	model, err := findModel(st)
+	if err != nil {
+		return nil, err
+	}
+
+	// check that all snaps from the model are constained by a validation set
+	// and check that all snaps from the model are valid in the validation sets.
+	if err := checkAllSnapsFromModelInValidationSets(model, valsets); err != nil {
+		return nil, err
+	}
+
+	// snaps required by validation sets must also be required by the model
+	if err := checkForRequiredSnapsNotRequiredInModel(model, valsets); err != nil {
+		return nil, err
+	}
+
+	tracker := snap.NewSelfContainedSetPrereqTracker()
+	offline := len(opts.LocalSnapPaths) > 0
+
+	var downloadTSS []*state.TaskSet
+	for _, sn := range model.AllSnaps() {
+		rev := revisions[sn.Name]
+
+		needsInstall, err := snapNeedsInstall(st, sn.Name, rev)
+		if err != nil {
+			return nil, err
+		}
+
+		if !needsInstall {
+			info, err := snapstate.CurrentInfo(st, sn.Name)
+			if err != nil {
+				return nil, err
+			}
+			tracker.Add(info)
+			continue
+		}
+
+		if offline {
+			info, err := offlineSnapInfo(sn, rev, opts)
+			if err != nil {
+				return nil, err
+			}
+			tracker.Add(info)
+			continue
+		}
+
+		const userID = 0
+		ts, info, err := snapstateDownload(context.TODO(), st, sn.Name, dirs.SnapBlobDir, &snapstate.RevisionOptions{
+			Channel:        sn.DefaultChannel,
+			Revision:       rev,
+			ValidationSets: valsets.Keys(),
+		}, userID, snapstate.Flags{}, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		tracker.Add(info)
+		downloadTSS = append(downloadTSS, ts)
+	}
+
+	warnings, errs := tracker.Check()
+	for _, w := range warnings {
+		logger.Noticef("create recovery system prerequisites warning: %v", w)
+	}
+
+	// TODO: use function from other branch
+	if len(errs) > 0 {
+		var builder strings.Builder
+		builder.WriteString("cannot create recovery system from model that is not self-contained:")
+
+		for _, err := range errs {
+			builder.WriteString("\n  - ")
+			builder.WriteString(err.Error())
+		}
+
+		return nil, errors.New(builder.String())
+	}
+
+	var snapsupTaskIDs []string
+	if len(downloadTSS) > 0 {
+		snapsupTaskIDs, err = extractSnapSetupTaskIDs(downloadTSS)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	chg = st.NewChange("create-recovery-system", fmt.Sprintf("Create new recovery system with label %q", label))
+	createTS, err := createRecoverySystemTasks(st, label, snapsupTaskIDs, opts)
 	if err != nil {
 		return nil, err
 	}
-	chg.AddAll(ts)
+
+	chg.AddAll(createTS)
+
+	for _, ts := range downloadTSS {
+		createTS.WaitAll(ts)
+		chg.AddAll(ts)
+	}
+
 	return chg, nil
 }
 
+func offlineSnapInfo(sn *asserts.ModelSnap, rev snap.Revision, opts CreateRecoverySystemOptions) (*snap.Info, error) {
+	index := -1
+	for i, si := range opts.LocalSnapSideInfos {
+		if sn.ID() == si.SnapID {
+			index = i
+			break
+		}
+	}
+	if index == -1 {
```

Comment:
we will land here also if the Model has Grade dangerous and somebody omitted snap-ids which is allowed, or some the snap.SideInfo doesn't have SnapID set.  We probably should leave a TODO here about supporting dangerous models, for now we might want to fail earlier if model.Grade is dangerous or if we can't do that, if either some model snap or one of the provided snap.SideInfos doesn't have a SnapdID?

Reply:
To support dangerous models, would it be sufficient to go off of the name as a back-up, in the situation that we don't have the snap's ID? That wouldn't be too hard of a change to make.

Reply:
yes, that would work. But as I tried to say we should accept SideInfos without SnapID only if the model is dangerous, that is the same rule as the seedwriter

Respectively ModelSnap might not have a snap-id set only if the model has grade dangerous